# Explore the Data
My own working notes to better understand what data I have access to that may be helpful in relating HIV-Associated Neurocognitive Disorder (HAND) with HIV sequence data.

## Outline
1. [Overview](#overview)  
2. [Sequence Data](#seqs)  
3. [Meta Data](#meta) 
4. [Preliminary Questions](#questions)

<a id='overview'></a>
## 1. Overview

### Ogishi and Yotsuyanagi 
Paper: https://retrovirology.biomedcentral.com/articles/10.1186/s12977-018-0401-x  
Data: https://github.com/masato-ogishi/HANDPrediction/tree/master/inst

Three data sets:
- HAND Database (data from the HAND Database)
- LANL (data from the Los Alamos National Laboratory HIV Database)
- Metadataset (data from the authors own curated set of sequences and metadata

Three types of files:
- __Sequence File__ C2V3C3_{DataSetName}_AA.fasta (amino acid sequence alignments for the C2V3C3 region of the HIV Envelope gene in fasta format)
- __Metadata File__ {DataSetName}.csv (metadata associated with each sequence {i.e. Hand status, patient id, accession number, publication, etc})
- __Sample Tissue File__ {DataSetName}_DesignSheet.csv (A mapping of Sample.Tissue { from the other .csv file} to Sample.Tissue.Category)

### Holman and Gabuzda
Paper: https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0049538  
Data: 
- (Validation seqs): https://github.com/veg/idepi/tree/master/examples/dementia/validation
- (MetaData): https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0049538#s5

A validation dataset that contains sequences from 10 individuals diagnosed with HAD, we have the sequences in the idepi codebase; just one file (HAD.fas) that contains the nucleotide sequences in fasta format.

Metadata for the training and testing datasets were available as excel files in the suplemental material of the paper. These are the two csv files saved in the HolmanAndGabuzda directory.

### idepi/Hepler
Paper: https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003842  
Data: https://github.com/veg/idepi/tree/master/examples/dementia

This is esentially the dataset used by Holman and Gabuzda (see above). The idepi project (by Hepler, et al.) used the Holman and Gabuzda data set for testing.

- The Training dataset: "HIVBrainSeqDBHepler.fasta" and "HAD.csv" (which just contains the HAD status)
- The Testing/validation dataset: "HAD.fas" (The fasta file for the 10 patients, all of whom were diagnosed with HAD

### HIVBrainSeqDB
Paper: https://aidsrestherapy.biomedcentral.com/articles/10.1186/1742-6405-7-43  
Data: The website has moved here http://hivbrainseqdb.dfci.harvard.edu/HIVSeqDB/; (http://www.HIVBrainSeqDB.org no longer works). 

I downloaded the entire HIV BrainSequence Database from http://hivbrainseqdb.dfci.harvard.edu/HIVSeqDB/ on 10-10-2018 and saved it in the data folder.

One data set with two files:
- HIVSeqDB.fasta (The nucleotide sequences in fasta format, from the HIV Brain Sequence Database)
- HIVSeqDB-MetaData.txv (the Metadata)

In [1]:
import pandas as pd
from Bio.SeqIO.FastaIO import SimpleFastaParser

<a id='seqs'></a>
## 2. Sequence Data (Fasta Files)

In [2]:
# Load in the fasta files and create a list of ids and a list of sequences.
Files = {}
Files['C2V3C3_HANDDatabase_AA.fasta'] = {'filePath': './OgishiAndYotsuyanagi/C2V3C3_HANDDatabase_AA.fasta'}
Files['C2V3C3_LANL_AA.fasta'] = {'filePath': './OgishiAndYotsuyanagi/C2V3C3_LANL_AA.fasta'}
Files['C2V3C3_Metadataset_AA.fasta'] = {'filePath': './OgishiAndYotsuyanagi/C2V3C3_Metadataset_AA.fasta'}
Files['HIVBrainSeqDB.fasta'] = {'filePath': './HIVBrainSeqDB/HIVSeqDB.fasta'}
Files['HIVBrainSeqDBHepler.fasta'] = {'filePath': './idepi/HIVBrainSeqDBHepler.fasta'}
Files['HAD.fas'] = {'filePath': './idepi/HAD.fas'}

for file in Files:
    with open(Files[file]["filePath"]) as fasta_file:
        ids = []
        sequences = []
        for title, sequence in SimpleFastaParser(fasta_file):
            ids.append(title.split(None, 1)[0])  # First word is ID
            sequences.append(sequence)
        Files[file]["ids"] = ids
        Files[file]["sequences"] = sequences

In [3]:
# Look at how many sequences and sites.
for file in Files:
    print(file)
    sequences = Files[file]["sequences"]
    seqLengths = []
    for seq in sequences:
        seqLengths.append(len(seq))
    
    uniqueSeqLengths = set(seqLengths)
    print('number of sites: ' + str(uniqueSeqLengths))
    print('number of sequences: ' + str(len(seqLengths)))
    print('\n')

C2V3C3_HANDDatabase_AA.fasta
number of sites: {189}
number of sequences: 1690


C2V3C3_LANL_AA.fasta
number of sites: {189}
number of sequences: 21609


C2V3C3_Metadataset_AA.fasta
number of sites: {189}
number of sequences: 2494


HIVBrainSeqDB.fasta
number of sites: {2560, 2562, 1541, 1030, 2565, 1032, 1033, 1545, 2571, 1020, 1544, 2574, 1039, 1036, 2577, 1042, 1044, 2583, 1048, 1047, 2586, 1023, 2589, 2592, 2595, 2601, 1066, 2607, 2608, 2610, 2613, 2614, 2616, 2617, 1081, 1082, 2625, 1093, 2119, 2634, 2635, 592, 2641, 595, 2644, 598, 2647, 601, 2650, 604, 607, 610, 613, 2665, 106, 619, 105, 622, 1650, 628, 631, 1153, 1164, 1167, 1170, 1173, 1179, 1182, 160, 161, 162, 163, 164, 1185, 1188, 170, 1197, 176, 1200, 1203, 181, 182, 1206, 2229, 1209, 1212, 189, 1215, 1218, 195, 1260, 198, 205, 206, 207, 1230, 1233, 210, 211, 1236, 213, 215, 216, 1239, 218, 219, 1242, 1245, 222, 1248, 225, 226, 1266, 1251, 230, 231, 1254, 1257, 234, 235, 236, 237, 238, 1263, 240, 241, 242, 243, 244, 245, 24

<a id='meta'></a>
## 3. Metadata (csv Files)
We won't print the HANDDatabase or LANL files from the Ogishi and Yotsuyanagi paper; they are very similar to the two files printed below for the Metadataset

In [4]:
# The Ogishi currated metadata dataset.
metadata = pd.read_csv('./OgishiAndYotsuyanagi/Metadataset.csv')
print (metadata.shape)
metadata.head()

(5386, 20)


Clinical.Status Accession  Reference_PublicationYear Reference_FirstAuthor  \
0            HAND  AF125810                       1999              Shapshak   
1            HAND  AF125811                       1999              Shapshak   
2            HAND  AF125812                       1999              Shapshak   
3            HAND  AF125813                       1999              Shapshak   
4            HAND  AF125814                       1999              Shapshak   

   Reference_PMID  LANL_Pub.id  LANL_SE.id  LANL_Patient.Id Patient.Sex  \
0        10381169         9244      231359            885.0           F   
1        10381169         9244      231358            885.0           F   
2        10381169         9244      231357            885.0           F   
3        10381169         9244      231356            885.0           F   
4        10381169         9244      231355            885.0           F   

   Number.of.patient.seqs      Georegion Coreceptor Sample.Tissue  \
0                    52.0  North America        NaN         brain   
1                    52.0  North America        NaN         brain   
2                    52.0  North America        NaN         brain   
3                    52.0  North America        NaN         brain   
4                    52.0  North America        NaN         brain   

  Culture.Method  Viral.load  CD4.count  CD8.count  Days.from.Infection  \
0            NaN         NaN        NaN        NaN                  NaN   
1            NaN         NaN        NaN        NaN                  NaN   
2            NaN         NaN        NaN        NaN                  NaN   
3            NaN         NaN        NaN        NaN                  NaN   
4            NaN         NaN        NaN        NaN                  NaN   

  Days.from.Seroconversion                                           Sequence  
0                     late  ACCCCACTCTGTGTTACTTTAAATTGCACTGATGCTAATTGGAGGA...  
1                     late  ACCCCACTCTGTGTTACTTTAAATTGCACTAATTGGGAGAATACTG...  
2                     late  ACCCCACTCTGTGTTACTTTAAATTGCACTAATTGGGAGAATATGT...  
3                     late  ACCCCACTCTGTGTTACTTTAAATTGCACTGATGCTAATTGGAGGA...  
4                     late  ACCCCACTCTGTGTTACTTTAAATTGCACTGATTGGAAGAATAGTA...

In [5]:
# The "DesignSheet" for the Ogishi currated metadata dataset.
metadataDesignSheet = pd.read_csv('./OgishiAndYotsuyanagi/Metadataset_SampleTissue_DesignSheet.csv')
print(metadataDesignSheet.shape)
metadataDesignSheet.head()

(20, 2)


Sample.Tissue Sample.Tissue.Category
0         blood                  Blood
1         brain                    CNS
2         colon                 Others
3           CSF                    CNS
4     diaphragm                 Others

In [6]:
# The HAD values from idepi associated with the Holman validation dataset.
HAD = pd.read_csv('./idepi/HAD.csv')
print(HAD.shape)
HAD.head()

(861, 2)


ID  HAD
0  DQ976387   50
1  DQ976386   50
2  DQ976385   50
3  DQ976384   50
4  DQ976383   50

In [7]:
# The metadata associated witht he Holman Training dataset.
HolmanTrain = pd.read_csv('./HolmanAndGabuzda/HolmanTrain.csv')
print(HolmanTrain.shape)
HolmanTrain.head()

(78, 12)


Publication    Patient  Sequence Count     Neurocognitive Diagnosis  \
0  Gatanaga, 1999 [1]  subject 4               6  HAD: severity not specified   
1    Lamers, 2009 [2]         DY              89                  HAD: severe   
2    Lamers, 2009 [2]         GA              52                  HAD: severe   
3    Lamers, 2009 [2]         CX             116  HAD: severity not specified   
4        Li, 1991 [3]         YU               2                  HAD: severe   

    Disease Status ART Treatment  CD4 Count  \
0  AIDS (deceased)           AZT        5.0   
1  AIDS (deceased)           yes        NaN   
2  AIDS (deceased)           yes      400.0   
3  AIDS (deceased)       unknown        NaN   
4  AIDS (deceased)           NaN        NaN   

   Viral Load (brain) copies/10e6 cells  \
0                                   NaN   
1                                   NaN   
2                                   NaN   
3                                   NaN   
4                                   NaN   

   Viral Load (plasma) copies/10e6 cells  \
0                                    NaN   
1                              7000000.0   
2                                    NaN   
3                                    NaN   
4                                    NaN   

   Viral Load (lymphoid) copies/10e6 cells  \
0                                      NaN   
1                                      NaN   
2                                      NaN   
3                                      NaN   
4                                      NaN   

                                       Tissue Source  Year of Death  
0                                                NaN            NaN  
1  University of California at San Francisco AIDS...            NaN  
2  University of California at San Francisco AIDS...            NaN  
3  University of California at San Francisco AIDS...            NaN  
4                                                NaN            NaN

In [8]:
# HIV Brain Sequence Database tsv file downloaded on 10-10-2018
HIVSeqDBMetadata = pd.DataFrame.from_csv('./HIVBrainSeqDB/HIVSeqDB-MetaData.tsv', sep='\t')
print(HIVSeqDBMetadata.shape)
print(list(HIVSeqDBMetadata.columns))
HIVSeqDBMetadata.head()

(2552, 26)
['Sample Tissue Code', 'Sample Tissue Class', 'Nucleic Acid Type', 'Sample Country', 'Sample City', 'Patient Age', 'Heath Status', 'Drug Naive', 'Antiretroviral Treatment', 'Viral Load Plasma', 'Viral Load Brain', 'Viral Load Lymphoid', 'CD4 Count', 'Neuropathological diagnosis', 'Neurocognitive diagnosis', 'Patient Code', 'Patient Sex', 'Patient Risk Factor', 'Patient Year of Death', 'Publication', 'PubMed Id', 'Publication Title', 'Sequence Start', 'Sequence Stop', 'Accession', 'Unnamed: 26']


/Users/ryanvelazquez/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:2: FutureWarning: from_csv is deprecated. Please use read_csv(...) instead. Note that some of the default arguments are different, so please refer to the documentation for from_csv when changing your function calls
  


Sample Tissue Code Sample Tissue Class Nucleic Acid Type  \
Sample Tissue Name                                                            
Spleen                       FMA:7196  Blood and Lymphoid               DNA   
Spleen                       FMA:7196  Blood and Lymphoid               DNA   
Spleen                       FMA:7196  Blood and Lymphoid               DNA   
Lymph node                   FMA:5034  Blood and Lymphoid               DNA   
Lymph node                   FMA:5034  Blood and Lymphoid               DNA   

                   Sample Country Sample City Patient Age     Heath Status  \
Sample Tissue Name                                                           
Spleen                         US     Chicago          39  AIDS (deceased)   
Spleen                         US     Chicago          39  AIDS (deceased)   
Spleen                         US     Chicago          39  AIDS (deceased)   
Lymph node                     US     Chicago          39  AIDS (deceased)   
Lymph node                     US     Chicago          39  AIDS (deceased)   

                   Drug Naive Antiretroviral Treatment  Viral Load Plasma  \
Sample Tissue Name                                                          
Spleen                     no                      AZT           194190.0   
Spleen                     no                      AZT           194190.0   
Spleen                     no                      AZT           194190.0   
Lymph node                 no                      AZT           194190.0   
Lymph node                 no                      AZT           194190.0   

                       ...      Patient Sex  Patient Risk Factor  \
Sample Tissue Name     ...                                         
Spleen                 ...                M      male homosexual   
Spleen                 ...                M      male homosexual   
Spleen                 ...                M      male homosexual   
Lymph node             ...                M      male homosexual   
Lymph node             ...                M      male homosexual   

                    Patient Year of Death   Publication PubMed Id  \
Sample Tissue Name                                                  
Spleen                             1993.0  Thomas, 2007  17084877   
Spleen                             1993.0  Thomas, 2007  17084877   
Spleen                             1993.0  Thomas, 2007  17084877   
Lymph node                         1993.0  Thomas, 2007  17084877   
Lymph node                         1993.0  Thomas, 2007  17084877   

                                                    Publication Title  \
Sample Tissue Name                                                      
Spleen              Macrophage entry mediated by HIV Envs from bra...   
Spleen              Macrophage entry mediated by HIV Envs from bra...   
Spleen              Macrophage entry mediated by HIV Envs from bra...   
Lymph node          Macrophage entry mediated by HIV Envs from bra...   
Lymph node          Macrophage entry mediated by HIV Envs from bra...   

                   Sequence Start Sequence Stop  Accession Unnamed: 26  
Sample Tissue Name                                                      
Spleen                       6225          8795   DQ976393         NaN  
Spleen                       6225          8795   DQ976392         NaN  
Spleen                       6225          8795   DQ976391         NaN  
Lymph node                   6225          8795   DQ976390         NaN  
Lymph node                   6225          8795   DQ976389         NaN  

[5 rows x 26 columns]

## 4. Some Preliminary Questions About the Data

### a. Why are there more rows in the metadata.csv then there are sequences (5,386 vs 2,494)... the paper says 2,492 sequences?

In [9]:
print('There are ',len(metadata.Accession.unique()), ' unique sequences in the metadata.csv file')
print('There are ',len(Files['C2V3C3_Metadataset_AA.fasta']["sequences"]), ' sequences in the Metadataset fasta file')

# I'm not sure of the reason for this but we can easily widdle down the metadata.csv file to only the sequences in the fasta file.
numberOfSequencesInTheFastaWithARowInTheCSV = 0
for index, row in metadata.iterrows():
    if row['Accession'] in Files['C2V3C3_Metadataset_AA.fasta']["ids"]:
        numberOfSequencesInTheFastaWithARowInTheCSV +=1
print('There are ', numberOfSequencesInTheFastaWithARowInTheCSV, 'sequences in the fasta that have a matching row in the csv')

There are  5386  unique sequences in the metadata.csv file
There are  2494  sequences in the Metadataset fasta file
There are  2494 sequences in the fasta that have a matching row in the csv


So... 
- There are over twice as many rows in the metadata.csv file than in the metadata.fasta file.
- All the sequences in the metadata.fasta file have a corresponding row in the csv file
- I'm not sure what the extra rows are...

### b. How many sequences are shared between files?
We won't look at the LANL file for now because we'll assume many of the sequences in other files are in there.

In [10]:
# Assign the lists of sequences to some short variables for easy access.
seqs={
    'LANLOgishiAA': Files['C2V3C3_LANL_AA.fasta']['ids'],
    'metaOgishiAA': Files['C2V3C3_Metadataset_AA.fasta']['ids'],
    'metaOgishi': list(metadata['Accession']),
    'HANDdbOgishi': Files['C2V3C3_HANDDatabase_AA.fasta']['ids'],
    'BraindbHepler': Files['HIVBrainSeqDBHepler.fasta']['ids'],
    'BrainDB': Files['HIVBrainSeqDB.fasta']['ids'],
    'valHolman': Files['HAD.fas']['ids']
}
# Create a matrix of overlaps
seqNumberMatrix = []
for key in seqs:
    LANLOverlap = len(set(seqs[key]).intersection(seqs['LANLOgishiAA']))
    metaOverlapAA = len(set(seqs[key]).intersection(seqs['metaOgishiAA']))
    metaOverlap = len(set(seqs[key]).intersection(seqs['metaOgishi']))
    HANDdbOgishiOverlap = len(set(seqs[key]).intersection(seqs['HANDdbOgishi']))
    BraindbHeplerOverlap = len(set(seqs[key]).intersection(seqs['BraindbHepler']))
    BrainDBOverlap = len(set(seqs[key]).intersection(seqs['BrainDB']))
    valHolmanOverlap = len(set(seqs[key]).intersection(seqs['valHolman']))
    
    seqNumberMatrix.append([LANLOverlap, metaOverlapAA, metaOverlap, HANDdbOgishiOverlap, BraindbHeplerOverlap, BrainDBOverlap, valHolmanOverlap])
# Create and print pandas DF
sharedSeqs = pd.DataFrame(seqNumberMatrix, columns = list(seqs.keys()), index = list(seqs.keys()))
print('Number of shared sequences')
sharedSeqs

Number of shared sequences


LANLOgishiAA  metaOgishiAA  metaOgishi  HANDdbOgishi  \
LANLOgishiAA          21609          1197        1568           697   
metaOgishiAA           1197          2494        2494          1449   
metaOgishi             1568          2494        5386          1449   
HANDdbOgishi            697          1449        1449          1690   
BraindbHepler           217           503         939           483   
BrainDB                 424           915        1917           912   
valHolman                 0             0           0             0   

               BraindbHepler  BrainDB  valHolman  
LANLOgishiAA             217      424          0  
metaOgishiAA             503      915          0  
metaOgishi               939     1917          0  
HANDdbOgishi             483      912          0  
BraindbHepler           1085     1085          0  
BrainDB                 1085     2552          0  
valHolman                  0        0         75

### c. How many sequences are there in total?
(excluding LANL)

In [11]:
print('excluding sequences from Ogishi without AA data')
len(set(seqs['metaOgishiAA'] + seqs['HANDdbOgishi'] + seqs['BraindbHepler'] + seqs['BrainDB'] + seqs['valHolman']))

excluding sequences from Ogishi without AA data


4352

In [12]:
print('including sequences from Ogishi without AA data')
len(set(seqs['metaOgishi'] + seqs['metaOgishiAA'] + seqs['HANDdbOgishi'] + seqs['BraindbHepler'] + seqs['BrainDB'] + seqs['valHolman']))

including sequences from Ogishi without AA data


6242

### d. How many patients are there in total?

This question (as well as others, including number 4 above) are probably best answered by trying to assemble a complete dataset with associated metadata. See the "Munge the Data" notebook.

In [13]:
# I started taking a look at this with:
patientIds = list(metadata['LANL_Patient.Id'])
patientIdsNoNAN = list(filter(lambda x: x >= 0, set(patientIds)))
print('pobably some where between ', len(set(patientIdsNoNAN)), ' and ',len(set(patientIds)), ' patients.')


pobably some where between  183  and  211  patients.
